File: sample/rs/15_wave_interference_loop.rs
mod py_runtime;
pub use crate::py_runtime::{math, pytra, time};
use crate::py_runtime::*;

use crate::time::perf_counter;
use crate::pytra::utils::gif::grayscale_palette;
use crate::pytra::utils::gif::save_gif;

// 15: Sample that renders wave interference animation and writes a GIF.

fn run_15_wave_interference_loop() {
    let w = 320;
    let h = 240;
    let frames_n = 96;
    let out_path = ("sample/out/15_wave_interference_loop.gif").to_string();
    
    let start = perf_counter();
    let mut frames: Vec<Vec<u8>> = vec![];
    
    let mut t: i64 = 0;
    while t < frames_n {
        let mut frame = vec![0u8; (w * h) as usize];
        let phase = ((t) as f64) * 0.12;
        let mut y: i64 = 0;
        while y < h {
            let row_base = y * w;
            let mut x: i64 = 0;
            while x < w {
                let dx = x - 160;
                let dy = y - 120;
                let v = math::sin((((x) as f64) + ((t) as f64) * 1.5) * 0.045) + math::sin((((y) as f64) - ((t) as f64) * 1.2) * 0.04) + math::sin((((x + y)) as f64) * 0.02 + phase) + math::sin(math::sqrt(dx * dx + dy * dy) * 0.08 - phase * 1.3);
                let mut c = (((v + 4.0) * (255.0 / 8.0)) as i64);
                if c < 0 {
                    c = 0;
                }
                if c > 255 {
                    c = 255;
                }
                let __idx_i64_1 = ((row_base + x) as i64);
                let __idx_2 = if __idx_i64_1 < 0 { (frame.len() as i64 + __idx_i64_1) as usize } else { __idx_i64_1 as usize };
                frame[__idx_2] = ((c) as u8);
                x += 1;
            }
            y += 1;
        }
        frames.push((frame).clone());
        t += 1;
    }
    save_gif(&(out_path), w, h, &(frames), &(grayscale_palette()), 4, 0);
    let elapsed = perf_counter() - start;
    println!("{} {}", ("output:").to_string(), out_path);
    println!("{} {}", ("frames:").to_string(), frames_n);
    println!("{} {}", ("elapsed_sec:").to_string(), elapsed);
}

fn main() {
    run_15_wave_interference_loop();
}

File: sample/ts/15_wave_interference_loop.ts
import * as math from "./pytra/std/math.js";
import { perf_counter } from "./pytra/std/time.js";
import { grayscale_palette } from "./pytra/utils/gif.js";
import { save_gif } from "./pytra/utils/gif.js";

// 15: Sample that renders wave interference animation and writes a GIF.

function run_15_wave_interference_loop() {
    let w = 320;
    let h = 240;
    let frames_n = 96;
    let out_path = "sample/out/15_wave_interference_loop.gif";
    
    let start = perf_counter();
    let frames = [];
    
    for (let t = 0; t < frames_n; t += 1) {
        let frame = (typeof (w * h) === "number" ? new Array(Math.max(0, Math.trunc(Number((w * h))))).fill(0) : (Array.isArray((w * h)) ? (w * h).slice() : Array.from((w * h))));
        let phase = t * 0.12;
        for (let y = 0; y < h; y += 1) {
            let row_base = y * w;
            for (let x = 0; x < w; x += 1) {
                let dx = x - 160;
                let dy = y - 120;
                let v = math.sin((x + t * 1.5) * 0.045) + math.sin((y - t * 1.2) * 0.04) + math.sin((x + y) * 0.02 + phase) + math.sin(math.sqrt(dx * dx + dy * dy) * 0.08 - phase * 1.3);
                let c = Math.trunc(Number((v + 4.0) * (255.0 / 8.0)));
                if (c < 0) {
                    c = 0;
                }
                if (c > 255) {
                    c = 255;
                }
                frame[(((row_base + x) < 0) ? ((frame).length + (row_base + x)) : (row_base + x))] = c;
            }
        }
        frames.push((Array.isArray((frame)) ? (frame).slice() : Array.from((frame))));
    }
    save_gif(out_path, w, h, frames, grayscale_palette());
    let elapsed = perf_counter() - start;
    console.log("output:", out_path);
    console.log("frames:", frames_n);
    console.log("elapsed_sec:", elapsed);
}

run_15_wave_interference_loop();
